### DISCUSSION TITLE
Search for campsites with specific attributes

### DISCUSSION THREAD

**Discussion by javagenki** [juftin/camply#63] [Ideas]
Thank again for the extremely helpful project. You’ve made finding and booking campsites so much easier.

Another thought I had, for a future revision, is to be have an optional configuration file that would allow a user to filter campsites based on certain attributes. For example, we booked a campsite at one campground where the competition is particularly fierce and didn’t notice during the feeding frenzy that the tent area was almost nonexistent - small tent size area was mentioned in the campsite details. At another time we found that our larger family car barely fit into the parking space provided for the site - smaller parking space was mentioned in the campsite details. We also have a dog…etc. We could have spent the time looking over each campsite’s details but when you have less than a minute - less than a few seconds for some sites - you can’t take in all of the details before you have to grab a site.

I think these details are exposed via the rec.gov API. I was thinking a separate YAML configuration file with your campsite preferences would both cut down on the number of irrelevant campsites returned in searches and would give you the comfort in knowing that the campsites returned will work for your situation when you need to make a split-second decision.

**Comment by juftin** [comment]
I'm already capturing the data - just not using it. What I'm not exactly sure about is a way to standardize the responses across different campgrounds to make it useful for lots of different setups.  This would be a useful way to dig in and find some patterns in the RIDB PermittedEquipment and Attributes Schema across a bunch of different campsites (note that the example campsite `98388` has a ton of data - I've also noticed responses with none or different data)
```python
import logging
from typing import List
from pprint import pprint
from camply.providers import RecreationDotGov
from camply.containers.api_responses import CampsiteResponse, _CampsiteAttribute, _CampsiteEquipment

logging.basicConfig(level=logging.DEBUG)

campsite_finder = RecreationDotGov()
campsite: CampsiteResponse = campsite_finder.get_campsite_by_id(campsite_id=98388)
attributes: List[_CampsiteAttribute] = campsite.ATTRIBUTES
permitted_equipment: List[_CampsiteEquipment] = campsite.PERMITTEDEQUIPMENT

pprint(attributes)
pprint(permitted_equipment)
```
```python
[_CampsiteAttribute(AttributeName='Food Locker', AttributeValue='Y'),
 _CampsiteAttribute(AttributeName='Picnic Table', AttributeValue='Y'),
 _CampsiteAttribute(AttributeName='Quiet Area', AttributeValue='Y'),
 _CampsiteAttribute(AttributeName='Site Access', AttributeValue='Drive-In'),
 _CampsiteAttribute(AttributeName='Shade', AttributeValue='Yes'),
 _CampsiteAttribute(AttributeName='Capacity/Size Rating', AttributeValue='Single'),
 _CampsiteAttribute(AttributeName='Checkin Time', AttributeValue='12:30 PM'),
 _CampsiteAttribute(AttributeName='Min Num of People', AttributeValue='1'),
 _CampsiteAttribute(AttributeName='Campfire Allowed', AttributeValue='Yes'),
 _CampsiteAttribute(AttributeName='Checkout Time', AttributeValue='12:00 PM'),
 _CampsiteAttribute(AttributeName='Pets Allowed', AttributeValue='Yes'),
 _CampsiteAttribute(AttributeName='Max Num of People', AttributeValue='8'),
 _CampsiteAttribute(AttributeName='Location Rating', AttributeValue='Good'),
 _CampsiteAttribute(AttributeName='IS EQUIPMENT MANDATORY', AttributeValue='true'),
 _CampsiteAttribute(AttributeName='Driveway Length', AttributeValue='60'),
 _CampsiteAttribute(AttributeName='Max Vehicle Length', AttributeValue='25'),
 _CampsiteAttribute(AttributeName='Tent Pad Width', AttributeValue='12'),
 _CampsiteAttribute(AttributeName='Driveway Entry', AttributeValue='Pull-Through'),
 _CampsiteAttribute(AttributeName='Driveway Surface', AttributeValue='Gravel'),
 _CampsiteAttribute(AttributeName='Driveway Grade', AttributeValue='Slight'),
 _CampsiteAttribute(AttributeName='Max Num of Vehicles', AttributeValue='1'),
 _CampsiteAttribute(AttributeName='Site Height/Overhead Clearance', AttributeValue='12'),
 _CampsiteAttribute(AttributeName='Tent Pad Length', AttributeValue='10'),
 _CampsiteAttribute(AttributeName='Min Num of Vehicles', AttributeValue='0'),
 _CampsiteAttribute(AttributeName='Site Width', AttributeValue='11')]


[_CampsiteEquipment(EquipmentName='Tent', MaxLength=25.0),
 _CampsiteEquipment(EquipmentName='RV', MaxLength=25.0),
 _CampsiteEquipment(EquipmentName='Trailer', MaxLength=25.0)]
```

**Comment by javagenki** [comment]
I’ve noticed that too - the data in the attributes seems to be all over the place. I’m not sure if this would work but could it be setup where you fill in the search parameters (YAML or command-line) and only those parameters are filtered - no other attributes are filtered? If the search parameter is matched then only those sites are returned? You’d probably miss a lot of valid sites this way if that attribute doesn’t exist or the data was bad but at least you’d know that the site is what you want for sure. In our case, we definitely need to know that ‘Pets Allowed’ = ‘Yes’.

BTW, your app has already helped to find some great campsites this summer so a big thank you from our family!

**Comment by ethan-leba** [comment]
Hi @juftin, I'd be interested in implementing attribute filtering if there's a clear consensus on how this can/should be implemented. 

> What I'm not exactly sure about is a way to standardize the responses across different campgrounds to make it useful for lots of different setups.

When you say that the data is unstandardized, do you mean across providers (rec.gov vs Yellowstone, etc) or across campgrounds on a single provider? Either way, I think using the non-standardized data would be a good intermediate solution for now.

What I'm not sure about though is how the filtering could be implemented in a flexible way. My current thought would be to allow users to input a python expression that's eval'ed for truthiness to filter the campsites after they've been retrieved. We could bind a `CAMPSITE` variable for eval, and so it could look something like `--post-process 'CAMPSITE.attrs["Pets Allowed"] == "Yes"'`.

**Comment by juftin** [comment]
I was primarily talking about non-standardization across recreation.gov campgrounds there. Looking directly at a dump of the database, there 329 different attribute names (I'll post the popular ones below) across all campgrounds. 

I like the API you suggested there. In the past I've done something like this by leveraging the pandas [df.eval](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.eval.html) syntax. I've never messed around much with `eval` before, pandas made it nice and safe which I liked. 

<details>
  <summary>All Attributes</summary>
  
| AttributeName | count |
| :--- | :--- |
| Checkin Time | 110181 |
| Checkout Time | 110121 |
| Max Num of People | 108080 |
| Max Num of Vehicles | 106157 |
| IS EQUIPMENT MANDATORY | 102402 |
| Pets Allowed | 100646 |
| Min Num of People | 99710 |
| Driveway Surface | 97195 |
| Capacity/Size Rating | 96084 |
| Max Vehicle Length | 95527 |
| Driveway Entry | 94447 |
| Campfire Allowed | 83536 |
| Shade | 81909 |
| Placed on Map | 75750 |
| Picnic Table | 62511 |
| Driveway Length | 47446 |
| Electricity Hookup | 41022 |
| Site Access | 40269 |
| Fire Pit | 39826 |
| Water Hookup | 34166 |
| Site Length | 30167 |
| Min Num of Vehicles | 23320 |
| Driveway Grade | 22285 |
| Site Width | 21946 |
| Location Rating | 20994 |
| Site Rating | 20921 |
| GRILLS | 20558 |
| Condition Rating | 19370 |
| CAMPFIRE RINGS | 17672 |
| BBQ | 15603 |
| Proximity to Water | 15025 |
| Tent Pad Length | 13768 |
| Tent Pad | 13524 |
| PICNIC TABLES | 13343 |
| Tent Pad Width | 13247 |
| Grills/Fire Ring | 12883 |
| Site Height/Overhead Clearance | 9988 |
| Base Number of People | 9456 |
| Base Number of Vehicles | 8871 |
| Hike In Distance to Site | 7253 |
| Double Driveway | 7166 |
| Accessibility | 7122 |
| Quiet Area | 6673 |
| FOOD STORAGE LOCKER | 5559 |
| LANTERN POSTS | 5261 |
| FLUSH TOILETS | 4928 |
| DRINKING WATER | 4610 |
| Food Locker | 4209 |
| PAVED ROADS | 4165 |
| Sewer Hookup | 4018 |
| Lantern Pole | 3702 |
| Accessible Occupant Message | 3385 |
| Privacy | 3357 |
| DUMP STATION | 3330 |
| TENT PADS | 3327 |
| ELECTRIC HOOKUPS | 3038 |
| HOST | 2893 |
| Shower/Bath Type | 2788 |
| WATER SPIGOT | 2667 |
| PAVED PARKING | 2641 |
| VAULT TOILETS | 2622 |
| SHOWERS | 2534 |
| WATER HOOKUPS | 2205 |
| BOAT RAMP | 2174 |
| TRASH COLLECTION | 2029 |
| Bed Type | 1973 |
| PARKING AREA | 1856 |
| ENTRANCE STATION | 1822 |
| Full Hookup | 1785 |
| Lean To/Shelter | 1720 |
| LAKE ACCESS | 1686 |
| ACCESSIBLE FLUSH TOILETS | 1648 |
| TRAILHEADS | 1637 |
| Max Num of Horses | 1544 |
| PAVED SITES | 1524 |
| EMERGENCY PHONE | 1480 |
| FLUSH TOILETS \(SEASONAL\) | 1372 |
| Num of Beds | 1362 |
| SELF PAY STATION | 1358 |
| CELL PHONE SERVICE | 1342 |
| ACCESSIBLE SHOWERS | 1279 |
| ACCESSIBLE CAMPSITES | 1276 |
| RECYCLING | 1274 |
| BEACH ACCESS | 1255 |
| PLAYGROUND | 1200 |
| ACCESSIBLE PARKING | 1185 |
| AMPHITHEATER | 1139 |
| PICNIC AREA | 1117 |
| Num of Rooms | 1102 |
| Platform | 1088 |
| Num of Bedrooms | 1037 |
| ACCESSIBLE DRINKING WATER | 1016 |
| DAY USE AREA | 1002 |
| ACCESSIBLE VAULT TOILETS | 994 |
| GENERAL STORE | 987 |
| DRINKING WATER \(PEAK SEASON\) | 968 |
| BOAT DOCK | 939 |
| RV PARKING | 905 |
| BOAT TRAILER PARKING | 899 |
| PICNIC SHELTERS | 865 |
| TABLE & BENCHES | 859 |
| WATER \(SEASONAL\) | 843 |
| CAMPFIRE CIRCLES | 841 |
| TABLES | 803 |
| FIREWOOD | 759 |
| ICE | 754 |
| ACCESSIBLE GRILLS | 706 |
| GIFT SHOP | 657 |
| UTILITY SINKS | 633 |
| VISITOR CENTER | 612 |
| RESTAURANT | 609 |
| Satellite TV Access | 596 |
| COIN SHOWERS | 582 |
| ELECTRICITY | 578 |
| SHOWERS \(SEASONAL\) | 566 |
| DRINKING WATER \(HAND PUMP\) | 558 |
| PIT TOILETS | 542 |
| RIVER ACCESS | 528 |
| DOCK | 513 |
| ACCESSIBLE SITES | 501 |
| ACCESSIBLE BOAT RAMP | 482 |
| TRAILHEAD PARKING | 457 |
| FIREWOOD VENDER | 455 |
| ACCESSIBLE PIT TOILETS | 429 |
| CONCESSIONS | 426 |
| ACCESSIBLE PICNIC AREA | 419 |
| Horse Stall/Corral | 413 |
| WIRELESS INTERNET | 406 |
| MARINA | 402 |
| WATERFRONT SITES | 401 |
| CREEK ACCESS | 396 |
| CAMPFIRE PROGRAMS | 387 |
| ACCESSIBLE PICNIC SHELTERS | 359 |
| EMERGENCY SERVICES | 357 |
| SCENIC OVERLOOKS | 345 |
| ACCESSIBLE PLAYGROUND | 332 |
| NON-POTABLE WATER | 324 |
| BOAT SLIPS | 312 |
| RANGER STATION | 298 |
| INTERPRETIVE TRAILS | 294 |
| ACCESSIBLE TRAILS | 284 |
| LAUNDRY FACILITIES | 272 |
| OFF-ROAD VEHICLE TRAILS | 272 |
| ACCESSIBLE BOCK DOCK | 268 |
| ATM | 256 |
| Horse Hitching Post | 255 |
| ACCESSIBLE WALKWAYS | 246 |
| Non Smoking | 241 |
| FUEL AVAILABLE | 240 |
| FISHING LICENSES | 237 |
| PAVED PATHWAYS | 229 |
| CORRALS | 226 |
| FISH CLEANING STATIONS | 226 |
| AMPHITHEATHER | 223 |
| LIGHTS | 219 |
| HORSEBACK RIDING TRAILS | 210 |
| Porch | 210 |
| PAY PHONE | 205 |
| ACCESSIBLE FISHING DOCK | 202 |
| PROPANE | 185 |
| Cabin Heating | 174 |
| Furnished | 174 |
| FIREPLACE | 173 |
| CHURCH | 171 |
| WASHER & DRYER | 171 |
| CATEGORY | 153 |
| MAP NUMBER | 152 |
| Toilet | 152 |
| Internet Access | 150 |
| WOOD STOVE | 150 |
| BOARDWALK | 147 |
| Generator Free | 142 |
| BED\(S\) WITH MATTRESS | 139 |
| BUNKBED\(S\) | 137 |
| CAVE TOURS | 136 |
| HORSE CORRALS | 135 |
| BOAT STORAGE | 133 |
| FISHING PIER | 133 |
| Lighting | 131 |
| BROOM & DUST PAN | 128 |
| BOAT RENTALS | 124 |
| CAMPING SUPPLIES | 124 |
| FISHING DOCK | 119 |
| COT\(S\) | 112 |
| GEOLOGICAL ATTRACTIONS | 110 |
| EDUCATIONAL PROGRAMS | 105 |
| Stove/Oven | 104 |
| Wood Heating | 103 |
| KENNEL | 102 |
| OUTHOUSE | 101 |
| Woodstove/Fireplace | 101 |
| Hot Water | 97 |
| TABLE & CHAIRS | 96 |
| SECURITY LIGHTS | 93 |
| Deck | 91 |
| PROPANE STOVE | 88 |
| Living Room | 84 |
| HITCHING RACKS | 79 |
| HORSESHOE PIT | 78 |
| KAYAK RENTALS | 77 |
| COOKING AREA | 76 |
| HEATING STOVE | 76 |
| MOTOR BOAT RENTALS | 74 |
| FIRE EXTINGUISHER | 71 |
| COOKING UTENSILS | 70 |
| COOKING POTS | 67 |
| ICE MACHINE | 67 |
| Stove/Fridge | 67 |
| FISHING GUIDE | 66 |
| VOLLEYBALL COURTS | 65 |
| PONTOON RENTAL | 64 |
| Cable TV Access | 60 |
| Propane Heating | 59 |
| PORTABLE TOILETS | 58 |
| SLEEPING SHELTER | 57 |
| Fridge | 56 |
| Fridge with Freezer | 56 |
| SHUTTLE BUS SERVICE | 56 |
| KITCHEN | 55 |
| SMOKE ALARM | 55 |
| BASKETBALL COURTS | 54 |
| LANTERN | 54 |
| HORSE STAGING AREA | 53 |
| Electric Heating | 52 |
| HISTORIC SITES | 52 |
| PROPANE OVEN | 52 |
| Clothes-Line | 51 |
| Microwave | 51 |
| Group Allowed | 50 |
| BUNKBEDS\(S\) WITH MATTRESSES | 48 |
| Oil Heating | 47 |
| Supplies | 46 |
| PROPANE HEAT | 44 |
| PROPANE LIGHTS | 44 |
| MOTEL | 40 |
| REFRIGERATOR | 40 |
| CLEANING SUPPLIES | 34 |
| BAIT SHOP | 33 |
| PLAYGROIUND | 33 |
| COFFEE POT | 32 |
| RV STORAGE | 32 |
| ACCESSIBLE SCENIC OVERLOOK | 31 |
| AXE | 31 |
| Air Conditioning | 29 |
| BED\(S\) | 29 |
| FOOT STORAGE LOCKER | 28 |
| HORSE STALLS | 28 |
| Manure Disposal | 28 |
| Fax Hookup | 26 |
| COUCHES | 25 |
| FIRE RING | 25 |
| LIFEGUARDS | 25 |
| PROPANE REFRIDGERATOR | 25 |
| TABLE | 25 |
| CLOTHES DRYER | 24 |
| FUTON\(S\) | 24 |
| WIND SHELTERS | 24 |
| OVEN | 23 |
| MOTORCYCLE TRAILS | 22 |
| DINING AREA | 21 |
| Gas Heating | 19 |
| OUTDOOR COOLER BOX | 19 |
| ELECTRIC STOVE | 18 |
| BUCKET & SHOVEL | 17 |
| Telephone Hookup | 16 |
| MATTRESS PAD\(S\) | 15 |
| EVENT AREA | 14 |
| GROUP COOKING FACILITIES | 13 |
| HEATER | 13 |
| SAW | 13 |
| TWIN BED\(S\) | 13 |
| BEAR POLE | 12 |
| FITNESS TRAILS | 12 |
| KITCHEN \(NO WATER\) | 12 |
| LOOKOUT TOWER | 12 |
| TENNIS COURTS | 12 |
| SOFTBALL FIELDS | 11 |
| HAS RENTALS | 10 |
| BOAT MAINTENANCE | 9 |
| WATER TROUGHS | 9 |
| BASEBALL FIELDS | 8 |
| FISH HATCHERY | 8 |
| DVD PLAYER | 7 |
| Telephone | 7 |
| CLEANING EQUIPMENT | 6 |
| ELECTRIC HOOKUPS 50 AMP | 6 |
| FIRST AID KIT | 6 |
| MOP | 6 |
| SNOW SHOVEL | 6 |
| BATHTUB | 5 |
| BAY ACCESS | 5 |
| BOAT INCLUDED | 5 |
| Cabin Electricity | 5 |
| Clothes-Dryer | 5 |
| Clothes-Washer | 5 |
| PROPANE GRILLS | 5 |
| PROPANE LANTERN \(FUEL NOT PROVIDED\) | 5 |
| COMMUNITY PARK | 4 |
| FENCED PASTURES | 4 |
| GROCERY STORE | 4 |
| Radio | 4 |
| BUTTERFLY GARDEN | 3 |
| COLEMAN STOVE \(FUEL NOT PROVIDED\) | 3 |
| HORSE WATERING STATION | 3 |
| SOCCER FIELDS | 3 |
| STOCK WATERING TANK | 3 |
| TWIN BED PLATFORM | 3 |
| ANCHORAGE | 2 |
| Dish-Washer | 2 |
| OFF-ROAD VEHICLE RENTALS | 2 |
| PROPANE FIREPLACE | 2 |
| SMALL REFRIDGERATOR | 2 |
| ARCHEOLOGICAL SITES | 1 |
| BATTERY LIGHTS | 1 |
| BOAT TRIPS | 1 |
| CANOE RENTALS | 1 |
| CLOTHESLINE | 1 |
| COLEMAN LANTERN \(FUEL NOT PROVIDED\) | 1 |
| DVD | 1 |
| FIRST AID STATION | 1 |
| LIGHTHOUSE | 1 |
| MEAT SHED | 1 |
| Maximum Group Capacity | 1 |
| OUTDOOR SHOWERS | 1 |
| PROPANE LANTERN | 1 |
| RAIN CATCHMENT BARREL | 1 |
| RV SUPPLIES | 1 |
| SHOOTING RANGE | 1 |
| Television | 1 |

</details>

**Reply by juftin** [reply]
recreation.gov by far gives us the most amount of attributes, some providers give none at all. If we could document some popular filtering use cases that would be super useful

**Reply by juftin** [reply]
In case you're interested in playing with the data, here's an old project that creates a SQLite db from RIDB. It'll create db file @ `~/.local/pipx/venvs/ridbpy/lib/python3.7/site-packages/ridbpy/data/ridb.db`

```shell
pipx install "ridbpy @ git+https://github.com/juftin/ridbpy@main" --python $(which python3.7)
ridbpy database raw fetch
ridbpy database raw ingest
open ~/.local/pipx/venvs/ridbpy/lib/python3.7/site-packages/ridbpy/data/
```

**Comment by juftin** [comment]
This discussion relates to https://github.com/juftin/camply/issues/86
